Makefile

```diff
@@ -48,4 +48,37 @@ submodules:
 
 # We must ensure up is installed in tool cache prior to build as including the k8s_tools machinery prior to the xpkg
 # machinery sets UP to point to tool cache.
-build.init: $(UP)
\ No newline at end of file
+build.init: $(UP)
+
+# ====================================================================================
+# End to End Testing
+
+KIND_VERSION = v0.16.0
+KIND_CLUSTER_NAME ?= uptest
+PROVIDER_GCP_VERSION ?= v0.15.0
+PROVIDER_HELM_VERSION ?= v0.12.0
+
+controlplane.up: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) setting up controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 || $(KIND) create cluster --name=$(KIND_CLUSTER_NAME)
+	@$(KUBECTL) -n upbound-system get cm universal-crossplane-config >/dev/null 2>&1 || $(UP) uxp install
+	@$(KUBECTL) -n upbound-system wait deploy crossplane --for condition=Available --timeout=120s
+	@$(KUBECTL) -n upbound-system create secret docker-registry package-pull-secret --docker-server=xpkg.upbound.io --docker-username=${UPBOUND_DOCKER_USERNAME} --docker-password=${UPBOUND_DOCKER_PASSWORD} -o yaml --dry-run=client | $(KUBECTL) apply -f -
+	@$(KUBECTL) get provider.pkg upbound-provider-gcp > /dev/null 2>&1 || $(UP) ctp provider install upbound/provider-gcp:$(PROVIDER_GCP_VERSION) --package-pull-secrets=package-pull-secret
+	@$(KUBECTL) get provider.pkg crossplane-contrib-provider-helm > /dev/null 2>&1 || $(UP) ctp provider install crossplane-contrib/provider-helm:$(PROVIDER_HELM_VERSION)
+	@$(KUBECTL) wait provider.pkg upbound-provider-gcp --for condition=Healthy --timeout=120s
+	@$(KUBECTL) wait provider.pkg crossplane-contrib-provider-helm --for condition=Healthy --timeout=120s
+	@$(OK) setting up controlplane
+
+controlplane.down: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) deleting controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 && $(KIND) delete cluster --name=$(KIND_CLUSTER_NAME)
+	@$(OK) deleting controlplane
+
+uptest-local: $(UP) $(KUBECTL) $(KUTTL)
+	@$(INFO) running automated tests
+	@$(KUBECTL) apply -R -f package/cluster
+	@KUBECTL=$(KUBECTL) KUTTL=$(KUTTL) uptest --claim-or-composite --example-list=examples/cluster-claim.yaml --default-timeout=2400 || $(FAIL)
```

Comment:
TODO(turkenh): This line needs to be replaced with available `uptest` binary.

Reply:
Should we add `uptest` to https://github.com/upbound/build?

Reply:
@turkenh where can I get the correct `uptest` binary version for local `uptest-local` testing?

Reply:
It is now part of build submodule, you should be able to checkout the PR and run `make e2e` after setting up your creds with UPTEST_GCP_CREDS env var.